packages/hint-compat-api/src/css.ts

```diff
@@ -1,29 +1,227 @@
- * Public
- * ------------------------------------------------------------------------------
+
+type ReportData = {
+    feature: string;
+    node: ChildNode;
+    unsupported: string[];
+};
+
+type ReportMap = Map<string, ReportData[] | null>;
+
+type Context = {
+    browsers: string[];
+    ignore: Set<string>;
+    report: (data: ReportData) => void;
+    walk: (ast: ContainerBase, context: Context) => void;
+};
+
+const getLocationFromNode = (node: ChildNode): ProblemLocation | undefined => {
+    const start = node.source && node.source.start;
+
+    return start && {
+        column: start.column - 1,
+        line: start.line - 1
+    };
+};
+
+const validateAtSupports = (node: AtRule, context: Context): void => {
+    const supported = filterSupports(node.params, context.browsers);
+
+    if (supported) {
+        context.walk(node, { ...context, browsers: supported });
+    }
+};
+
+const validateAtRule = (node: AtRule, context: Context): void => {
+    if (node.name === 'supports') {
+        validateAtSupports(node, context);
+
+        return;
+    }
+
+    const unsupported = getUnsupported({ rule: node.name }, context.browsers);
+
+    if (unsupported) {
+        context.report({ feature: `@${node.name}`, node, unsupported });
+    }
+
+    context.walk(node, context);
+};
+
+const validateDeclValue = (node: Declaration, context: Context) => {
+    const unsupported = getUnsupported({ property: node.prop, value: node.value }, context.browsers);
+
+    if (unsupported) {
+        context.report({ feature: `${node.prop}: ${node.value}`, node, unsupported });
+    }
+};
+
+const validateDecl = (node: Declaration, context: Context) => {
+    const property = node.prop;
+
+    if (context.ignore.has(property) || context.ignore.has(`${property}: ${node.value}`)) {
+        return;
+    }
+
+    const unsupported = getUnsupported({ property }, context.browsers);
+
+    if (unsupported) {
+        context.report({ feature: `${property}`, node, unsupported });
+    } else {
+        validateDeclValue(node, context);
+    }
+};
+
+const validateDeclInBlock = (node: Declaration, reports: ReportMap, context: Context): void => {
+    const prop = vendor.unprefixed(node.prop);
+
+    // If any matching decl was already fully supported, no need to evaluate further.
+    if (reports.get(prop) === null) {
+        return;
+    }
+
+    let reportData: ReportData | null = null;
+
+    const report = (data: ReportData) => {
+        reportData = data;
+    };
+
+    // Validate decl, caching the report (if any) in reportData.
+    validateDecl(node, { ...context, report });
+
+    if (reportData) {
+        if (!reports.has(prop)) {
+            reports.set(prop, []);
+        }
+
+        // Decl had a report, add it to the list to resolve later.
+        reports.get(prop)!.push(reportData);
+    } else {
+        // Decl was fully supported; no need to test other decls in this block.
```

Comment:
It took me a while to understand what was going on here with this comment. Maybe something like the following?

> Decl was fully supported; no need to test other decls for same property in this block so setting to `null` to `return` at the beginning of the function.